src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import linear_algebra.matrix.nonsingular_inverse
+
+/-!
+# Integer powers of square matrices
+
+In this file, we define integer power of matrices, relying on
+the nonsingular inverse definition for negative powers.
+
+## Implementation details
+
+The main definition is a direct recursive call on the integer inductive type,
+as provided by the `div_inv_monoid.gpow` default implementation.
+The lemma names are taken from `algebra.group_with_zero.power`.
+
+## Tags
+
+matrix inverse, matrix powers
+-/
+
+open_locale matrix
+
+namespace matrix
+
+variables {n' : Type*} [decidable_eq n'] [fintype n'] {K : Type*} [comm_ring K]
+
+lemma smul_pow {α R : Type*} [semiring α] [monoid R] [distrib_mul_action R α]
+  [is_scalar_tower R α α] [smul_comm_class R α α]
+  (k : R) (A : matrix n' n' α) (p : ℕ) :
+  (k • A) ^ p = k ^ p • A ^ p :=
+begin
+  induction p with p IH,
+  { simp },
+  { rw [pow_succ', IH, smul_mul_smul, ←pow_succ', ←pow_succ'] }
+end
+
+local notation `M` := matrix n' n' K
+
+noncomputable instance : div_inv_monoid M :=
+{ ..(show monoid M, by apply_instance),
+  ..(show has_inv M, by apply_instance) }
+
+section nat_pow
+
+@[simp] theorem inv_pow' (A : M) (n : ℕ) : (A⁻¹) ^ n = (A ^ n)⁻¹ :=
+begin
+  by_cases h : is_unit A.det,
```

Comment:
I see quite a few proofs in this PR that unfold the definition of `⁻¹` by doing a cases on `is_unit A.det`, and there's not much mathematically interesting going on in there. Do you have any suggestions for missing API around inverses? In particular, would some sort of "induction principle" for inverses help?

Something along the lines of:
```lean
def matrix.cases_on_inv (A : matrix n n R) (P : matrix n n R → Sort*)
  (hinv : ∀ B, A \con B = 1 → B \con A = 1 → P B)
  (hsingl : (∀ B, A \con B ≠ 1) → P 0) (hsingr : (∀ B, B \con A ≠ 1) → P 0) :
  P (A⁻¹) := sorry
```